src/main/java/com/powsybl/openloadflow/ac/MonitoringVoltageOuterLoop.java

```diff
@@ -0,0 +1,158 @@
+                        LOGGER.trace("Switch bus '{}' PQ -> PV with low targetV={}", controllerBus.getId(), newTargetV);
+                    }
+                }
+            }
+        }
+
+        Reports.reportPqToPvBuses(reporter, pqPvSwitchCount, 0);
+
+        return pqPvSwitchCount > 0;
+    }
+
+    private static double getBusV(LfBus bus) {
+        return bus.getVoltageControl().map(vc -> vc.getControlledBus().getV()).orElse(Double.NaN);
+    }
+
+    /**
+     * A PQ bus with a static var compensator monitoring voltage can be switched to PV in 2 cases:
+     *  - if the voltage of the controlled bus is higher than the high voltage threshold.
+     *  - if the voltage of the controlled bus is lower that the low voltage threshold.
+     */
+    private static void checkPqBusForVoltageLimits(LfBus controllerCapableBus, List<PqToPvBus> pqToPvBuses, Range<Double> voltageLimit) {
+        double busV = getBusV(controllerCapableBus);
+        if (busV > voltageLimit.getMaximum()) {
+            pqToPvBuses.add(new PqToPvBus(controllerCapableBus, VoltageLimitDirection.MAX));
+        }
+        if (busV < voltageLimit.getMinimum()) {
+            pqToPvBuses.add(new PqToPvBus(controllerCapableBus, VoltageLimitDirection.MIN));
+        }
+    }
+
+    private static Optional<LfStaticVarCompensator> findMonitoringSvc(LfBus bus) {
+        return bus.getGenerators().stream()
+                .filter(gen -> gen.getGeneratorControlType() == LfGenerator.GeneratorControlType.MONITORING_VOLTAGE)
+                .findFirst()
+                .map(LfStaticVarCompensator.class::cast);
+    }
+
+    private static Optional<Range<Double>> getControlledBusVoltageLimits(LfBus controllerCapableBus) {
+        return findMonitoringSvc(controllerCapableBus)
+                .flatMap(generator -> generator.getStandByAutomaton()
+                        .map(automaton -> Range.between(automaton.getLowVoltageThreshold(), automaton.getHighVoltageThreshold())));
+    }
+
+    private static double getSvcTargetV(LfBus bus, VoltageLimitDirection direction) {
+        return findMonitoringSvc(bus).flatMap(svc -> svc.getStandByAutomaton()
+            .map(automaton -> {
+                svc.setGeneratorControlType(LfGenerator.GeneratorControlType.VOLTAGE); // FIXME
+                if (direction == VoltageLimitDirection.MIN) {
+                    return automaton.getLowTargetV();
+                } else {
+                    return automaton.getHighTargetV();
+                }
+            })
+        )
+        .orElse(Double.NaN);
+    }
+
+    @Override
+    public OuterLoopStatus check(OuterLoopContext context, Reporter reporter) {
+        OuterLoopStatus status = OuterLoopStatus.STABLE;
+
+        List<PqToPvBus> pqToPvBuses = new ArrayList<>();
+        for (LfBus bus : context.getNetwork().getBuses()) {
+            if (bus.hasVoltageControllerCapability() && !bus.isDisabled()) {
+                if (!bus.hasGeneratorsWithSlope()) {
```

Comment:
Why do we need to check this? We can never have both a generator controlling reactive power and a SVC with a standby automaton?